### PR TITLE
[DS-2958] Automated documentation generation for REST API: Swagger + swagger UI

### PR DESCRIPTION
**Pull request by KevinVdV**
https://jira.duraspace.org/browse/DS-2958

Incomplete adding of swagger and swagger UI to the rest api of DSpace. The ItemResource & CollectionResource have been partially documented as proof of concept.
In order to run this locally some configuration will need to be added to the local.cfg file (it has already been added to the local.cfg.EXAMPLE):

# Sets the schemes for the API URLs (http, https).
rest.schemes=https
# Sets the host (including a port) for the API URLs. Does not include the schemes nor context root.
rest.host=localhost:8080
# Sets the context root for the API calls.
rest.setBasePath=/rest/api

Still todo:
- [ ] Pin the swagger UI version, do not depend on "master"
- [ ] Add DSpace color scheme & logo to the swagger UI
- [ ] Document the BitstreamResource endpoint
- [ ] Document the CollectionsResource endpoint
- [ ] Document the CollectionsResource endpoint
- [ ] Document the CommunitiesResource endpoint
- [ ] Document the FilteredCollectionsResource endpoint
- [ ] Document the FilteredItemsResource endpoint
- [ ] Document the FiltersResource endpoint
- [ ] Document the HandleResource endpoint
- [ ] Document the HierarchyResource endpoint
- [ ] Document the ItemsResource endpoint
- [ ] Document the MetadataRegistryResource endpoint
- [ ] Document the RestIndex endpoint
- [ ] Document the RestReports endpoint

### CODE REVIEW

**Comment by peterdietz** [comment]
Cool thing to add, thanks Kevin. One discussion point. Would we want to keep the GET /, RestIndex.sayHtmlHello method that outputs java string of our endpoints, or would we agree to have the swagger output be the index?

File: dspace/config/modules/rest.cfg [hunk 1]
```diff
@@ -4,6 +4,18 @@
...
 #---------------------------------------------------------------#
 
+# Sets the schemes for the API URLs (http, https).
+rest.schemes=${rest.schemes}
+
+# Sets the host (including a port) for the API URLs. Does not include the schemes nor context root.
+rest.host=${rest.host}
+
+# Sets the context root for the API calls.
+rest.setBasePath=${rest.setBasePath}
```
Comment (tdonohue) [inline dspace/config/modules/rest.cfg]: These settings cause DSpace to have a circular default reference (a bad practice). This is the *old* way (build.properties), and is no longer supported or recommended.

Instead, the default values should be placed in this `rest.cfg` file. They can always be overridden by copying the config into your local.cfg.

**Comment by KevinVdV** [comment]
@tdonohue Thank you for the tips on the configuration, I was unsure on how to proceed with these, added a new commit that will (hopefully) fix it.

**Comment by tdonohue** [comment]
Thanks @KevinVdV , the new commit looks correct (though I'm not sure if the default `rest.host` should be just `localhost`, or if it needs a port, i.e.`localhost:8080`).

Travis CI now seems to be only be complaining about license headers (i.e. the previously failing unit tests now succeed).

**Comment by KevinVdV** [comment]
@tdonohue I was in doubt about wether to put the host to localhost or localhost:8080, the thing is that the rest currently requires you to be on https, so adding port 8080 makes less sense.

@peterdietz Not sure if we need the "sayHtmlHello" method, the swagger gives you are more interactive way to navigate through all the rest methods, so I'm inclined towards dropping it.


Also marking this PR as in progress as it isn't done yet (see checklist above). If anybody wants to help out, more information on swagger annotations can be found here: https://github.com/swagger-api/swagger-core/wiki/Annotations-1.5.X (But let me know that you are working on it so our work doesn't collide).

**Comment by hardyoyo** [comment]
Any progress report on this? It looks very cool, but, if others are actively working on it, I'd prefer to not step on toes.

**Comment by KevinVdV** [comment]
@hardyoyo Currently having issues with the swagger UI & the way our DSpace rest is currently working.

I get the following javascript exception whenever I document a method using a community object:
client.js:1192 Uncaught RangeError: Maximum call stack size exceeded

This is because a community can contain another community and apperently swaggerUI can't handle this setup.

Trying to find out how best to resolve this.

**Comment by tdonohue** [comment]
Thanks for your Pull Request. However, I'm closing this PR as it conflicts with our upcoming DSpace 7 REST API work (taking place currently on our `rest7` feature branch, and moving over to `master` soon). 

This PR can still move forward through one of the following routes:
1. You can choose to rework this PR to align with the new 7.x REST API efforts. Keep in mind, this will be a rebuilt REST API, and not all features have been completed/rebuilt. But, we welcome other collaborators to help move the effort forward. See https://wiki.duraspace.org/display/DSPACE/DSpace+7+UI+Working+Group
2. Or, you can resubmit this PR to our `dspace-6_x` maintenance branch as a simple patch to our 6.x REST API.

Again, thanks for the work. We hope you consider joining us in the DSpace 7 REST API efforts, but we also appreciate any patches to the existing 6.x REST API you are willing to provide.

**Pull request closed without merging**